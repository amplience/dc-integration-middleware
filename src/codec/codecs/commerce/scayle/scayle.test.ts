import ScayleCodecType, { CodecConfig, ScayleCodec } from '.'
import { CodecPropertyConfig, CommerceCodec } from '../../core'
import nock, { Scope } from 'nock'
import {
	categorySingleResponse,
	productNotFoundResponse,
	productsFirstPageResponse,
	productsLastPageResponse,
	productsMultipleResponse,
	productsSingleResponse
} from './test/responses'

const config: CodecPropertyConfig<CodecConfig> = {
	access_token: 'test_access_token',
	shop_id: 'test_shop_id',
	tenant_space: 'dc-integration-middleware-test-demo',
	api_version: 'v1'
}

const assertIsProduct = (obj) => {
	expect(obj).toEqual(
		expect.objectContaining({
			id: expect.any(String),
			name: expect.any(String),
			slug: expect.any(String),
			categories: expect.any(Array),
			variants: expect.any(Array),
			longDescription: expect.any(String),
			shortDescription: expect.any(String)
		})
	)
}

const assertIsCategory = (obj) => {
	expect(obj).toEqual(
		expect.objectContaining({
			id: expect.any(String),
			name: expect.any(String),
			slug: expect.any(String),
			children: expect.any(Array),
			products: expect.any(Array),
			showInMenu: expect.any(Boolean)
		})
	)
}

describe('scayle integration', () => {
	let codec: CommerceCodec
	let nockScope: Scope

	beforeAll(async () => {
		nock.disableNetConnect()
	})

	afterAll(async () => {
		nock.enableNetConnect()
	})

	beforeEach(async () => {
		jest.resetAllMocks()

		nockScope = nock('https://dc-integration-middleware-test-demo.storefront.api.scayle.cloud/v1')
		codec = new ScayleCodec(config)
		await codec.init(new ScayleCodecType())
	})

	test('getProduct by ID', async () => {
		nockScope
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				ids: '1',
				page: '1',
				perPage: '20'
			})
			.reply(200, productsSingleResponse)

		const result = await codec.getProduct({
			id: '1'
		})

		expect(result.id).toEqual('1')
		assertIsProduct(result)
		expect(result).toMatchSnapshot()
	})

	test('getProduct by ID - not found', async () => {
		nockScope
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				ids: 'not_found',
				page: '1',
				perPage: '20'
			})
			.reply(200, productNotFoundResponse)

		const result = await codec.getProduct({
			id: 'not_found'
		})

		expect(result).toBeNull()
	})

	test('getProducts by IDs', async () => {
		nockScope
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				ids: '1,2',
				page: '1',
				perPage: '20'
			})
			.reply(200, productsMultipleResponse)

		const result = await codec.getProducts({
			productIds: '1,2'
		})

		expect(result.length).toEqual(2)
		assertIsProduct(result[0])
		expect(result[0].id).toEqual('1')
		assertIsProduct(result[1])
		expect(result[1].id).toEqual('2')
		expect(result).toMatchSnapshot()
	})

	test('getProducts by ID - partial results (one not found)', async () => {
		nockScope
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				ids: '1,not_found',
				page: '1',
				perPage: '20'
			})
			.reply(200, productsSingleResponse)

		const result = await codec.getProducts({
			productIds: '1,not_found'
		})

		expect(result.length).toEqual(2)
		assertIsProduct(result[0])
		expect(result[0].id).toEqual('1')
		expect(result[1]).toBeNull()
	})

	test('getProducts by keyword', async () => {
		nockScope
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				'filters[term]': 'shirt',
				page: '1',
				perPage: '20'
			})
			.reply(200, productsSingleResponse)

		const result = await codec.getProducts({
			keyword: 'shirt'
		})

		expect(result.length).toEqual(1)
		assertIsProduct(result[0])
		expect(result[0].id).toEqual('1')
		expect(result).toMatchSnapshot()
	})

	test('getProducts by keyword - paginated', async () => {
		nockScope
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				'filters[term]': 'shirt',
				page: '1',
				perPage: '2'
			})
			.reply(200, productsFirstPageResponse)
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				'filters[term]': 'shirt',
				page: '2',
				perPage: '2'
			})
			.reply(200, productsLastPageResponse)

		const result = await codec.getProducts({
			keyword: 'shirt',
			pageSize: 2
		})

		expect(result.length).toEqual(4)
		expect(result[0].id).toEqual('1')
		expect(result[1].id).toEqual('2')
		expect(result[2].id).toEqual('3')
		expect(result[3].id).toEqual('4')
		expect(result).toMatchSnapshot()
	})

	test('getProducts by category id', async () => {
		nockScope
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				'filters[category]': '1',
				page: '1',
				perPage: '20'
			})
			.reply(200, productsSingleResponse)

		const result = await codec.getProducts({
			category: {
				id: '1',
				name: 'test_category_name',
				slug: 'test_category_slug',
				children: [],
				products: [],
				showInMenu: true
			}
		})

		expect(result.length).toEqual(1)
		assertIsProduct(result[0])
		expect(result[0].id).toEqual('1')
		expect(result).toMatchSnapshot()
	})

	test('getProducts by categories - paginated', async () => {
		nockScope
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				'filters[category]': '1',
				page: '1',
				perPage: '2'
			})
			.reply(200, productsFirstPageResponse)
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				'filters[category]': '1',
				page: '2',
				perPage: '2'
			})
			.reply(200, productsLastPageResponse)

		const result = await codec.getProducts({
			category: {
				id: '1',
				name: 'test_category_name',
				slug: 'test_category_slug',
				children: [],
				products: [],
				showInMenu: true
			},
			pageSize: 2
		})

		expect(result.length).toEqual(4)
		expect(result[0].id).toEqual('1')
		expect(result[1].id).toEqual('2')
		expect(result[2].id).toEqual('3')
		expect(result[3].id).toEqual('4')
		expect(result).toMatchSnapshot()
	})

	test('getCategory by slug', async () => {
		nockScope
			.get('/categories')
			.query({
				shopId: 'test_shop_id'
			})
			.reply(200, categorySingleResponse)
			.get('/products')
			.query({
				shopId: 'test_shop_id',
				with: 'variants,images,priceRange,attributes:key(description|name|category)',
				'filters[category]': '1',
				page: '1',
				perPage: '20'
			})
			.reply(200, productsSingleResponse)

		const result = await codec.getCategory({ slug: 'maenner' })

		assertIsCategory(result)
		expect(result).toMatchSnapshot()
	})

	test('getCategoryTree', async () => {
		nockScope
			.get('/categories')
			.query({
				shopId: 'test_shop_id'
			})
			.reply(200, categorySingleResponse)
		const result = await codec.getCategoryTree({})

		expect(result.length).toEqual(1)
		assertIsCategory(result[0])
		expect(result).toMatchSnapshot()
	})
})
