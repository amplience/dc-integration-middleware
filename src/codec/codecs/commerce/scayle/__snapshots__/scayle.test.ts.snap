// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scayle integration getCategory by slug 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "id": "3",
          "name": "Shirts",
          "products": [],
          "showInMenu": true,
          "slug": "shirts",
        },
      ],
      "id": "2",
      "name": "Bekleidung",
      "products": [],
      "showInMenu": true,
      "slug": "bekleidung",
    },
    {
      "children": [],
      "id": "4",
      "name": "Hosen",
      "products": [],
      "showInMenu": true,
      "slug": "hosen",
    },
    {
      "children": [],
      "id": "5",
      "name": "Schuhe",
      "products": [],
      "showInMenu": true,
      "slug": "schuhe",
    },
  ],
  "id": "1",
  "name": "Männer",
  "products": [
    {
      "categories": [
        {
          "children": [],
          "id": "3",
          "image": undefined,
          "name": "Damen",
          "products": [],
          "showInMenu": true,
          "slug": undefined,
        },
      ],
      "id": "1",
      "longDescription": "",
      "name": "Leviz T-Shirt",
      "shortDescription": "",
      "slug": "LEV0074002000005",
      "variants": [
        {
          "attributes": undefined,
          "id": "31",
          "images": [
            {
              "url": "https://example/file/images/68bf124a0517bffaa5fe1dc7ac8707db.jpg?brightness=0.96&quality=75&trim=1&height=1067&width=800",
            },
          ],
          "listPrice": "€2,990.00",
          "salePrice": "€2,990.00",
          "sku": "LEV0074002000005_S",
        },
      ],
    },
  ],
  "showInMenu": true,
  "slug": "maenner",
}
`;

exports[`scayle integration getCategoryTree 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "3",
            "name": "Shirts",
            "products": [],
            "showInMenu": true,
            "slug": "shirts",
          },
        ],
        "id": "2",
        "name": "Bekleidung",
        "products": [],
        "showInMenu": true,
        "slug": "bekleidung",
      },
      {
        "children": [],
        "id": "4",
        "name": "Hosen",
        "products": [],
        "showInMenu": true,
        "slug": "hosen",
      },
      {
        "children": [],
        "id": "5",
        "name": "Schuhe",
        "products": [],
        "showInMenu": true,
        "slug": "schuhe",
      },
    ],
    "id": "1",
    "name": "Männer",
    "products": [],
    "showInMenu": true,
    "slug": "maenner",
  },
]
`;

exports[`scayle integration getProduct by ID 1`] = `
{
  "categories": [
    {
      "children": [],
      "id": "3",
      "image": undefined,
      "name": "Damen",
      "products": [],
      "showInMenu": true,
      "slug": undefined,
    },
  ],
  "id": "1",
  "longDescription": "",
  "name": "Leviz T-Shirt",
  "shortDescription": "",
  "slug": "LEV0074002000005",
  "variants": [
    {
      "attributes": undefined,
      "id": "31",
      "images": [
        {
          "url": "https://example/file/images/68bf124a0517bffaa5fe1dc7ac8707db.jpg?brightness=0.96&quality=75&trim=1&height=1067&width=800",
        },
      ],
      "listPrice": "€2,990.00",
      "salePrice": "€2,990.00",
      "sku": "LEV0074002000005_S",
    },
  ],
}
`;

exports[`scayle integration getProducts by IDs 1`] = `
[
  {
    "categories": [
      {
        "children": [],
        "id": "7",
        "image": undefined,
        "name": "Herren",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "1",
    "longDescription": "",
    "name": "Helfeger Shirt",
    "shortDescription": "",
    "slug": "THS0906008000001",
    "variants": [
      {
        "attributes": undefined,
        "id": "1",
        "images": [
          {
            "url": "https://example/file/cd81a5d37e4241686586812d9e179553?bg=F4F4F5&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€3,990.00",
        "salePrice": "€3,990.00",
        "sku": "THS0906008000001_S",
      },
    ],
  },
  {
    "categories": [
      {
        "children": [],
        "id": "7",
        "image": undefined,
        "name": "Herren",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "2",
    "longDescription": "",
    "name": "Helfeger Shirt",
    "shortDescription": "",
    "slug": "THS0906006000001",
    "variants": [
      {
        "attributes": undefined,
        "id": "4",
        "images": [
          {
            "url": "https://example/file/8de1d6ebb091d7363a05bcf99a0095f3?quality=75&height=1067&width=800",
          },
        ],
        "listPrice": "€3,990.00",
        "salePrice": "€3,990.00",
        "sku": "THS0906006000001_S",
      },
    ],
  },
]
`;

exports[`scayle integration getProducts by categories - paginated 1`] = `
[
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "1",
    "longDescription": "",
    "name": "Liu Jo Sandale CAMELIA",
    "shortDescription": "",
    "slug": "LIJ0480001000004",
    "variants": [
      {
        "attributes": undefined,
        "id": "40",
        "images": [
          {
            "url": "https://example/file/images/bc06c53f15b5a6634e10747378fc9eb7.png?bg=F4F4F5&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€32,990.00",
        "salePrice": "€32,990.00",
        "sku": "LIJ0480001000004_38",
      },
    ],
  },
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "2",
    "longDescription": "",
    "name": "Leviz T-Shirt",
    "shortDescription": "",
    "slug": "LEV0031009000002",
    "variants": [
      {
        "attributes": undefined,
        "id": "37",
        "images": [
          {
            "url": "https://example/file/38ee8cd16ca9f4e7912ec4df24d1c753?bg=F4F4F5&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€2,990.00",
        "salePrice": "€2,990.00",
        "sku": "LEV0031009000002_S",
      },
    ],
  },
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "3",
    "longDescription": "",
    "name": "Leviz T-Shirt",
    "shortDescription": "",
    "slug": "LEV0031001002000",
    "variants": [
      {
        "attributes": undefined,
        "id": "34",
        "images": [
          {
            "url": "https://example/file/6074e8f1a0b1cd81885a1769e5d0cfa1?bg=F4F4F5&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€2,990.00",
        "salePrice": "€2,990.00",
        "sku": "LEV0031001002000_S",
      },
    ],
  },
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "4",
    "longDescription": "",
    "name": "Leviz T-Shirt",
    "shortDescription": "",
    "slug": "LEV0074002000005",
    "variants": [
      {
        "attributes": undefined,
        "id": "31",
        "images": [
          {
            "url": "https://example/file/images/68bf124a0517bffaa5fe1dc7ac8707db.jpg?brightness=0.96&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€2,990.00",
        "salePrice": "€2,990.00",
        "sku": "LEV0074002000005_S",
      },
    ],
  },
]
`;

exports[`scayle integration getProducts by category id 1`] = `
[
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "1",
    "longDescription": "",
    "name": "Leviz T-Shirt",
    "shortDescription": "",
    "slug": "LEV0074002000005",
    "variants": [
      {
        "attributes": undefined,
        "id": "31",
        "images": [
          {
            "url": "https://example/file/images/68bf124a0517bffaa5fe1dc7ac8707db.jpg?brightness=0.96&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€2,990.00",
        "salePrice": "€2,990.00",
        "sku": "LEV0074002000005_S",
      },
    ],
  },
]
`;

exports[`scayle integration getProducts by keyword - paginated 1`] = `
[
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "1",
    "longDescription": "",
    "name": "Liu Jo Sandale CAMELIA",
    "shortDescription": "",
    "slug": "LIJ0480001000004",
    "variants": [
      {
        "attributes": undefined,
        "id": "40",
        "images": [
          {
            "url": "https://example/file/images/bc06c53f15b5a6634e10747378fc9eb7.png?bg=F4F4F5&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€32,990.00",
        "salePrice": "€32,990.00",
        "sku": "LIJ0480001000004_38",
      },
    ],
  },
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "2",
    "longDescription": "",
    "name": "Leviz T-Shirt",
    "shortDescription": "",
    "slug": "LEV0031009000002",
    "variants": [
      {
        "attributes": undefined,
        "id": "37",
        "images": [
          {
            "url": "https://example/file/38ee8cd16ca9f4e7912ec4df24d1c753?bg=F4F4F5&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€2,990.00",
        "salePrice": "€2,990.00",
        "sku": "LEV0031009000002_S",
      },
    ],
  },
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "3",
    "longDescription": "",
    "name": "Leviz T-Shirt",
    "shortDescription": "",
    "slug": "LEV0031001002000",
    "variants": [
      {
        "attributes": undefined,
        "id": "34",
        "images": [
          {
            "url": "https://example/file/6074e8f1a0b1cd81885a1769e5d0cfa1?bg=F4F4F5&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€2,990.00",
        "salePrice": "€2,990.00",
        "sku": "LEV0031001002000_S",
      },
    ],
  },
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "4",
    "longDescription": "",
    "name": "Leviz T-Shirt",
    "shortDescription": "",
    "slug": "LEV0074002000005",
    "variants": [
      {
        "attributes": undefined,
        "id": "31",
        "images": [
          {
            "url": "https://example/file/images/68bf124a0517bffaa5fe1dc7ac8707db.jpg?brightness=0.96&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€2,990.00",
        "salePrice": "€2,990.00",
        "sku": "LEV0074002000005_S",
      },
    ],
  },
]
`;

exports[`scayle integration getProducts by keyword 1`] = `
[
  {
    "categories": [
      {
        "children": [],
        "id": "3",
        "image": undefined,
        "name": "Damen",
        "products": [],
        "showInMenu": true,
        "slug": undefined,
      },
    ],
    "id": "1",
    "longDescription": "",
    "name": "Leviz T-Shirt",
    "shortDescription": "",
    "slug": "LEV0074002000005",
    "variants": [
      {
        "attributes": undefined,
        "id": "31",
        "images": [
          {
            "url": "https://example/file/images/68bf124a0517bffaa5fe1dc7ac8707db.jpg?brightness=0.96&quality=75&trim=1&height=1067&width=800",
          },
        ],
        "listPrice": "€2,990.00",
        "salePrice": "€2,990.00",
        "sku": "LEV0074002000005_S",
      },
    ],
  },
]
`;
